### ISSUE TITLE
Vagrant up & Docker dependencies

### ISSUE THREAD

**Issue by p-l-** [mitchellh/vagrant#4980]
Currently, when Vagrant is used with the Docker provider and there are dependencies (links between containers), Vagrant attempts to start all of them simultaneously.

Since there is no way to force the `--no-parallel` behavior from the `Vagrantfile` (see issue #4843), we need to ask users, as a workaround, to either set the environment variable `VAGRANT_NO_PARALLEL` before running `vagrant up` or to use the `--no-parallel` flag.

Beside the need for extra parameters than the simple `vagrant up`, this is not an ideal workaround because we loose the benefits of parallel starts for containers once dependencies are met.

For example, say you have one database server and two web servers that depend on the database server but do not depend on each other. Using `--no-parallel` will start the containers one after the other, instead of starting (ideally) the database server alone and then the two web servers in parallel.

Would it be feasible that Vagrant uses the `.link()` method of the Docker provider configuration to create a start order?

**Comment by schmunk42** [comment]
:+1: 

Linking and automatic usage of no-parallel is a very import feature, since it leads to a lot of user errors, when missing the correct options. 

[fig](http://www.fig.sh) - aka docker-compose in the future - handles the startup process quite well and should be a reference for implementation.

**Comment by jure** [comment]
Definitely a :+1:, it fails so frequently that it detracts from the beauty of `vagrant up`. Some potential workarounds are mentioned here: https://github.com/docker/fig/issues/374 

For example, you could have a script in the Dockerfile that waits until its link is ready, and only then continues, e.g.:

```
until nc -z postgres 5432; do
    echo "$(date) - waiting for postgres..."
    sleep 1
done
```

**Comment by bezerker** [comment]
Any updates on this?  This is only thing that keeps my users confused at times.

**Comment by ncoghlan** [comment]
I'm hitting the original problem described in the closed issue #4843 when trying to use Vagrant to smoothly automate setting up a local multi-VM Kubernetes cluster. I'd like to use an existing multihost Ansible playbook, but that means wait to provision the master until after all the node VMs have started.

Not having any way in the Vagrant file to say either "run all provisioning in series" or "provision the master VM after all the node VMs" means that "vagrant up" may or may not work depending on when the master gets provisioned relative to the nodes.

**Comment by cweagans** [comment]
FWIW, putting `ENV['VAGRANT_NO_PARALLEL'] = 'yes'` in my Vagrantfile seems to work okay in my setup.

**Comment by LongLiveCHIEF** [comment]
I have an idea that's based on @crazyh suggestion.  His suggestion:

```
config.vm.define "db2" , order: 1
config.vm.define "frontend" , order: 2
config.vm.define "proxy" , order: 3
```

Perhaps instead of an `order` method, `define` could support `before` and `after`. Examples:

Three machines, first machine needs to be built synchronously, remaining two in parrallel:

```
config.vm.define "docker-host", primary:true do |docker|
   ....
end

config.vm.define "docker-redis", after: "docker-host" do |redis|
  ...
end

config.vm.define "docker-nginx", after: "docker-host" do |nginx|
   ...
end
```

Another example, 3 vm's, and the last one depends on all the others to be built first, with the first two being build in parallel:


```
config.vm.define "docker-host", before: "before-app-dashboard" do |database|
   ....
end

config.vm.define "web-app", before: "app-monitor-dashboard" do |web|
  ...
end

config.vm.define "app-monitor-dashboard" do |dashboard|
   ...
end
```

There are a few major advantages to this I can think of right away:

- you can support the use of `primary: true` to enable auto-provisioning of dependent VM's
- you can sequence groups of parallel vm builds, maintaining speed in creation where it's needed
- you can use the output of other VM's (via synced folders) to iterate on your VM builds

**Comment by LongLiveCHIEF** [comment]
Here's an [example of a setup](https://github.com/HackerHappyHour/impact/blob/e3988996c049914eb3b4be2e6442525bd24cfe57/Vagrantfile) where I'd want to use these features right now.

**Comment by cweagans** [comment]
Definitely +1 to not having a manually defined order. `before` and `after` should be more than adequate to determine which boxes can be brought up together and in what order.

**Comment by sethvargo** [comment]
Hi everyone - doing triage here. 

This looks to be a duplicate of https://github.com/mitchellh/vagrant/issues/4026, which provides for more than just Docker VM dependencies. Let's track that instead. Thanks! 😄
